ethexe/processor/src/handling/run.rs

```diff
@@ -16,178 +16,315 @@
+                    .instantiate()
+                    .expect("Failed to instantiate executor");
+                let program_id = *program_id;
+                let state_hash = *state_hash;
 
-    loop {
-        // Send tasks to process programs in workers, until all queues are empty.
+                let _ = join_set.spawn_blocking(move || {
+                    let (jn, new_state_hash) =
+                        run_runtime(db, &mut executor, program_id, state_hash);
+                    (task_num, program_id, new_state_hash, jn)
+                });
+            }
 
-        let mut no_more_to_do = true;
-        for index in (0..in_block_transitions.states_amount()).step_by(virtual_threads) {
-            let result_receivers = one_batch(index, &task_senders, in_block_transitions).await;
+            let bucket_size = bucket.len();
+            let mut handler = DeterministicJournalHandler::new(bucket_size);
 
-            let mut super_journal = vec![];
-            for (program_id, receiver) in result_receivers.into_iter() {
-                let (journal, dispatch_origin) = receiver.await.unwrap();
-                if !journal.is_empty() {
-                    super_journal.push((
-                        program_id,
-                        dispatch_origin.expect("origin should be set for non-empty journal"),
-                        journal,
-                    ));
-                    no_more_to_do = false;
-                }
-            }
+            while let Some((task_num, program_id, new_state_hash, program_journals)) = join_set
+                .join_next()
+                .await
+                .transpose()
+                .expect("Failed to join task")
+            {
+                // State was updated during journal handling inside the runtime (allocations, pages)
+                in_block_transitions.modify(program_id, |state, _| {
+                    state.hash = new_state_hash;
+                });
 
-            for (program_id, dispatch_origin, journal) in super_journal {
-                let mut handler = JournalHandler {
-                    program_id,
-                    controller: TransitionController {
-                        transitions: in_block_transitions,
-                        storage: &db,
-                    },
-                    dispatch_origin,
-                };
-                core_processor::handle_journal(journal, &mut handler);
+                handler.set_journal_part(task_num, program_id, program_journals);
+                handler.try_handle_journal_part(
+                    &db,
+                    in_block_transitions,
+                    &mut no_message_processed,
+                );
             }
         }
 
-        if no_more_to_do {
+        if no_message_processed {
             break;
         }
     }
+}
 
-    for handle in handles {
-        handle.abort();
+fn split_to_buckets(
+    virtual_threads: usize,
+    states: &Vec<(ActorId, H256, usize)>,
```

Comment:
what about `impl Iterator<(...)>`

Reply:
also reference to vec is not welcome in rust

Reply:
also, also as I can see - this is simple sort by queue size method

Reply:
> also, also as I can see - this is simple sort by queue size method

Please note that this is not exactly sorting (sorting usually `n*log(n)` this one is `O(n)`), but rather partitioning into subsets (buckets) of programs with approximately similar queue sizes. That's actually where the name split_to_buckets comes from. So, it doesn't seem entirely accurate to call it sorting.

Reply:
Fixed